### DISCUSSION TITLE
Could there be a section for documenting approaches to testing plugins

### DISCUSSION THREAD

**Discussion by claremacrae** [marcusolsson/obsidian-plugin-docs#39] [Ideas]
Hi, thank you for some awesome docs pages!

I’m an experienced developer with a background in depending on automatic tests to speed up development…

So as I start to learn JS, to start to write plugins, I’m naturally looking for guidance on writing automated tests as I go…

And I wondered whether you would be interested in seeing any techniques and tools documented here…

**Comment by claremacrae** [comment]
It’s early days, but I’ve started collecting links as I look around for info:

### Testing Plugins

* End-to-end testing
    * For plugin developers: Write test cases that interact with a running obsidian instance!
        * [For plugin developers: Write test cases that interact with a running obsidian instance! - Share & showcase - Obsidian Forum](https://forum.obsidian.md/t/for-plugin-developers-write-test-cases-that-interact-with-a-running-obsidian-instance/16574)
        * [GitHub - trashhalo/obsidian-plugin-e2e-test: sample repo with obsidian plugin e2e tests](https://github.com/trashhalo/obsidian-plugin-e2e-test)
        * This has a [GitHub Action](https://github.com/trashhalo/obsidian-plugin-e2e-test/actions) that runs the tests
* Plugins that have tests
    * dataview
        * [obsidian-dataview/src/test at master · blacksmithgu/obsidian-dataview · GitHub](https://github.com/blacksmithgu/obsidian-dataview/tree/master/src/test)
        * This has a [GitHub Action](https://github.com/blacksmithgu/obsidian-dataview/actions) that runs the tests
* Requests for testability
    * [MockObisidan testing utility · Issue #37 · obsidian-tools/obsidian-tools · GitHub](https://github.com/obsidian-tools/obsidian-tools/issues/37)

**Comment by claremacrae** [comment]
The sorts of content I’m thinking of include:

- Separating the “business logic” of your plugin (like, the code that generates text that is inserted in to the editor) from your UI code - so that you can test the behaviour for a selection in inputs, without needing to control the UI
- How to automate running of your tests in GitHub Actions
- Maybe testing handling of settings

**Comment by marcusolsson** [comment]
Documentation on how to test your plugins would be a wonderful addition to the docs! You're way ahead of me here, so I'd definitely appreciate the help in distilling the resources to something that would be useful to most plugin authors!

I didn't know about Spectron, but it seems really useful here! I might play around with it a bit. I'll read up on the resources you posted as well!

**Comment by claremacrae** [comment]
Thank you! 

One aim might be for there to be a fork of [obsidianmd/obsidian-sample-plugin ](https://github.com/obsidianmd/obsidian-sample-plugin) that has basic tests included, with the GitHub Action set up - so that people who wanted tests could start by forking that and extending it…

**Comment by claremacrae** [comment]
Just to note: I’ve asked on OMG Discord’s #plugin-dev channel if there’s anyone else who can add info…

https://discord.com/channels/686053708261228577/840286264964022302/896790151353753610

**Comment by claremacrae** [comment]
Hi Marcus, I’ve got a ton of info about testing plugins now. My writing skills are not up to  yours, but I really want to share what I have learned. 

It may start off long and then get divided up. 

I offer it here or in the community hub. 

I would sketch out the headings and get feedback on the structure and location before doing the writing. 

I’d prefer it to go here, as more developers are like to land here. 

What do you think, would you accept it as a contribution?

**Comment by marcusolsson** [comment]
Would definitely be up for that idea! Maybe open a pull request that we could collaborate on?

**Comment by claremacrae** [comment]
OK so my talk on this topic is in 6 days' time - so, current status: going through notes, trying to form some structure out of them.

Will have to learn the formatting tool you use here.
